[release/6.0][wasm] Add support for native relinking after Build, and AOT after publish

radical opened the pull request:
Wasm app build can run in two scenarios:

1. For build, eg. when running the app in VS, or `dotnet build foo.csproj`
2. For Publish, eg, when publishing the app to a folder, or Azure

A dotnet wasm app has some native wasm files (`dotnet.wasm`, and `dotnet.js`). How these files are obtained, or generated:

- Build
    - with no native libraries referenced (AOT setting is ignored here)
        - files from the runtime pack are used as-is
    - with native libraries referenced
        - dotnet.wasm is relinked with the native libraries

- Publish
    - dotnet.wasm is relinked with the native libraries, and updated pinvoke/icalls from the trimmed assemblies
    - if `RunAOTCompilation=true`, then the relinking includes AOT'ed assemblies

## `Build`

Implementation:

- Target `WasmBuildApp`
- runs after `Build` by default
    - which can be disabled by `$(DisableAutoWasmBuildApp)`
    - or the run-after target can be set via `$(WasmBuildAppAfterThisTarget)`

- To run a custom target
    - *before* any of the wasm build targets, use `$(WasmBuildAppDependsOn)`, and prepend your target name to that
    - *after* any of the wasm build targets, use `AfterTargets="WasmBuildApp"` on that target
- Avoid depending on this target, because it is available only when the workload is installed. Use `$(WasmNativeWorkload)` to check if it is installed.

## `Publish`

Implementation:

- This part runs as a nested build using a `MSBuild` task, which means that the project gets reevaluated. So, if there were any changes made to items/properties in targets before this, then they won't be visible in the nested build.
- By default `WasmTriggerPublishApp` runs after the `Publish` target, and that triggers the nested build
    - The nested build runs `WasmNestedPublishApp`, which causes `Build`, and `Publish` targets to be run
    - Because this causes `Build` to be run again, if you have any targets that get triggered by that, then they will be running twice.
        - But the original *build* run, and this *publish* run can be differentiated using `$(WasmBuildingForPublish)`

- `WasmTriggerPublishApp` essentially just invokes the nested publish
    - This runs after `Publish`
        - which can be disabled by `$(DisableAutoWasmPublishApp)`
        - or the run-after target can be set via `$(WasmTriggerPublishAppAfterThisTarget)`

    - To influence the wasm build for publish, use `WasmNestedPublishApp`
        - To run a custom target before it, use `$(WasmNestedPublishAppDependsOn)`
        - to run a custom target *after* it, use `AfterTargets="WasmNestedPublishApp"`

    - If you want to *dependsOn* on this, then use `DependsOnTargets="WasmTriggerPublishApp"`

- Fixes dotnet#56783
- Fixes dotnet#53612
- Fixes dotnet#58973

- Also fixes building for AOT with trimming disabled
Fixes dotnet#53804

radical committed:
Fix helix work item name

radical committed:
fix test

radical committed:
get all the assemblies from the publishdir

radical committed:
AOTCompiler: Expand the paths used with MONO_PATH

radical committed:
Implement support for skipping assemblies for AOT

radical committed:
Update blazorwasm targets file with a locally patched copy

radical committed:
Cleanup

radical committed:
fix path for copying blazorwasm targets file

radical committed:
Fix tests on emsdk

radical committed:
Don't fixup runtime pack paths when not using workloads

fix build

really fix the build

radical committed:
rebased

radical committed:
[wasm] Fix building blazor projects with no workload installed

- And add extra checks for verifying that

radical committed:
Update wasm README.md

radical committed:
Add missing files

radical committed:
WasmApp.targets: remove hack for blazor, which is not needed now

radical committed:
add comments for the new properties

radical committed:
blazor-sdk: cleanup assemblies list before adding to it

radical committed:
Fix "Clean" for blazorwasm projects, for *Build*

Partially fixes https://github.com/dotnet/runtime/issues/58826

radical committed:
Update BlazorOverride.targets to 6.0

radical committed:
run blazor tests only with the workloads installed

radical committed:
[wasm] Rename tmp folder 'relink' to 'for-build'

Renames the temporary folder used for native files built after *Build*
target, from `relink` to `for-build`, to be consistent with publish.

radical committed:
[wasm] Fix support for AOT with trimming disabled

When publishing with AOT (`RunAOTCompilation=true`), the assemblies get
trimmed by the linker, and copied to a single `linked/` folder. Then
`MonoAOTCompiler` runs `mono-aot-cross` on those assemblies. Since, all
the assemblies have their dependencies in the same folder too,
`mono-aot-cross` can resolve them from there.

But when trimming is disabled (`PublishTrimmed=false`),
`MonoAOTCompiler` gets paths to assemblies from different folders. For
example, some might be from `obj`, nugets etc. Which means that
`mono-aot-cross` might not always be able to find dependencies for the
assemblies being AOTed.

To fix this, if all the input assemblies are not in the same directory,
then we copy them to a temporary one, and then run `mono-aot-cross` on
them.

Fixes #53804

radical committed:
WasmAppBuilder - cleanup

radical committed:
Bump sdk for workload testing to 6.0.100-rc.2.21459.15

radical committed:
disable  non-wasm builds

radical committed:
disable  non-wasm builds

radical committed:
disable  non-wasm builds

radical committed:
[wasm] Fix Publish with VS

When publishing with VS, it builds with `$(DeployOnBuild)=true`. That
causes `DotNetPublish` to run, which runs `Publish` after `Build`.

Wasm targets run `WasmTriggerPublishApp`, that then uses the MSBuild
task to invoke `WasmNestedPublishApp, which depends on `Publish`, which
depends on `Build`. And that causes `DotNetPublish` again, which results
in:

```
sdk/6.0.100-rc.2.21459.15/Sdks/Microsoft.NET.Sdk.Publish/targets/CopyTargets/Microsoft.NET.Sdk.Publish.CopyFiles.targets(80,5):

error MSB4006: There is a circular dependency in the target dependency graph involving target "Publish".
Since "_CopyAspNetCoreFilesToIntermediateOutputPath" has "DependsOn" dependence on "Publish", the circular is "Publish<-_CopyAspNetCoreFilesToIntermediateOutputPath<-CorePublish<-DotNetPublish<-_PublishBuildAlternative<-Publish".
```

Fix is to unset `DeployOnBuild` for the nested build
(`WasmNestedPublishApp`).

radical committed:
Revert "disable  non-wasm builds"

This reverts commit a93db63a49856130ac17558b0a95ffb24c780d4d.

radical committed:
Fix up blazorOverride.targets

radical committed:
Fix local aot builds for samples

radical committed:
Fix aot with no trimming for blazor projects

.. and correctly skip assemblies marked for non-aot.

radical commented:
wasm test failure is https://github.com/dotnet/runtime/issues/58202:

```
[10:07:48] fail: [out of order message from the browser]: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'ws://127.0.0.1:32993/WebSocket/EchoWebSocket.ashx?subprotocol=CustomProtocol' failed: Error during WebSocket handshake: Response must not include 'Sec-WebSocket-Protocol' header if not present in request: CustomProtocol
[10:07:48] fail: [out of order message from the browser]: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'wss://127.0.0.1:40283/WebSocket/EchoWebSocket.ashx?subprotocol=CustomProtocol' failed: Error during WebSocket handshake: Response must not include 'Sec-WebSocket-Protocol' header if not present in request: CustomProtocol
[10:07:48] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'ws://127.0.0.1:39461/' failed: WebSocket is closed before the connection is established.
[10:07:48] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'ws://24ec1fbf05574cd4b5be6f210fbbe53d/' failed: WebSocket is closed before the connection is established.
[10:07:49] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'ws://127.0.0.1:32993/WebSocket/EchoWebSocket.ashx?delay10sec' failed: WebSocket is closed before the connection is established.
[10:07:49] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'wss://127.0.0.1:40283/WebSocket/EchoWebSocket.ashx?delay10sec' failed: WebSocket is closed before the connection is established.
[10:07:50] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'ws://127.0.0.1:32993/WebSocket/EchoWebSocket.ashx?delay10sec' failed: WebSocket is closed before the connection is established.
[10:07:50] info: http://127.0.0.1:32993/dotnet.js 0 WebSocket connection to 'wss://127.0.0.1:40283/WebSocket/EchoWebSocket.ashx?delay10sec' failed: WebSocket is closed before the connection is established.
[10:08:37] fail: [FAIL] System.Net.WebSockets.Client.Tests.MemorySendReceiveTest.ReceiveAsync_MultipleOutstandingReceiveOperations_Throws(server: ws://127.0.0.1:32993/WebSocket/EchoWebSocket.ashx)
[10:08:37] info: Closed state when OperationCanceledException
[10:08:37] info: Expected: True
[10:08:37] info: Actual:   False
[10:08:37] info:    at System.Net.WebSockets.Client.Tests.SendReceiveTest.ReceiveAsync_MultipleOutstandingReceiveOperations_Throws(Uri server)
[10:08:37] info: --- End of stack trace from previous location ---
[10:12:05] info: Finished:    System.Net.WebSockets.Client.Tests.dll
```

radical committed:
Wasm.Build.Test: find differently cased publish dir

`Expected to find /datadisks/disk1/work/AF47099B/w/AABA0938/e/blz_deploy_on_build_Debug_True/bin/Debug/net6.0/publish/wwwroot/_framework/dotnet.wasm`

.. though the path from the logs:
`bin/Debug/net6.0/Publish/wwwroot/_framework/dotnet.wasm`

.. with the publish dir case being `publish` vs `Publish`, and so the
test fails on linux, as it is looking for `publish`.

radical committed:
Rename WasmBuildingForPublish->WasmBuildingForNestedPublish to be clearer

radical committed:
Update BlazorOverwrite.targets

1. Ensure that the PublishDir uses the same case as PublishUrl, to
   workaround a bug where we end up with two directories on linux:
   `Publish`, and `publish`.
   https://github.com/dotnet/sdk/issues/20923

2. Add all the `@(WasmNativeAssets)` as candidates, and not just
   `.wasm`

radical committed:
sample/iOS: Add new required IntermediateOutputPath parameter for MonoAOTCompiler

radical committed:
Fix more wasm.build.tests to work around blazor's Publish dir issue

radical committed:
Surface @(FileWrites) from nested publish to the main build

radical committed:
[wasm] Don't relink after build, when `$(DeployOnBuild)=true`

When publishing from VS, it builds the project with
`$(DeployOnBuild)=true`, in which case we can skip relinking after
Build, and instead do it only after publish.

Fixes https://github.com/dotnet/runtime/issues/58973

radical committed:
Revert "disable  non-wasm builds"

This reverts commit 1e67f3e0f430be3c2b1db924feebc06bc9cafb49.

jonathanpeppers commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -280,46 +287,69 @@ public override bool Execute()
+            throw new LogAsErrorException($"'{nameof(OutputType)}=Library' can not be used with '{nameof(UseStaticLinking)}=true'.");
+        }
+
+        return true;
```
You might change this to:
```suggestion
        return !Log.HasLoggedErrors;
```

Just in case something in here logged an error, and it still got here.

radical committed:
Merge remote-tracking branch 'origin/release/6.0' into build-publish-6.0

radical committed:
Update BlazorOverwrite.targets to work with rc2

radical committed:
[wasm] Disable a blazorwasm publish test

`DefaultTemplate_AOT_OnlyWithPublishCommandLine_Then_PublishNoAOT`

This test runs build thrice:
1. simple build (no relinking)
2. publish+aot
3. publish-no-aot (relinking)

This causes the publish folder to have more than one `dotnet*hash*js`
files, which breaks the test since we don't know which one is the
"current" one.

radical committed:
Bump sdk for workload testing to 6.0.100-rc.2.21463.12

radical committed:
Merge remote-tracking branch 'origin/release/6.0' into build-publish-6.0

radical committed:
Revert "Update BlazorOverwrite.targets to work with rc2"

This reverts commit 48085f2d0683948c1d53958e62820515f502e059.

- We want to revert this to a version that works with the currently
  available SDK, which doesn't have the ComputeBlazorPublishAsset fixes,
  or the new cache-busting changes.

radical committed:
Address review feedback from @jonathanpeppers

radical committed:
Revert "disable  non-wasm builds"

This reverts commit 3da2650b1125660b5c9ae5578b0fa1c485aa1425.
